jujugui/static/gui/src/app/app.js

```diff
@@ -856,6 +867,39 @@ YUI.add('juju-gui', function(Y) {
     },
 
     /**
+      Handles redirecting to a new model when attempting to log in.
+
+      @method _handleRedirectionRequired
+      @param {String} err The error message string.
+      @param {Object} resp The response value containing the available server
+        details.
+    */
+    _handleRedirectionRequired: function(err, resp) {
+      // If we have a response and we have a redirection required error
+      // then we need to modify the connection information and try again.
+      // Otherwise we just carry on not doing anything.
+      if (resp && (err && err.indexOf('redirection required') > -1)) {
+        // Loop through the available servers and find the public IP.
+        const server = resp.servers[0].filter(
```

Comment:
I think we should just provide the list of servers to callbacks not the response itself. Also, what do you think about keeping the signature of loginWithMacaroon as is (just returning an error, not a response) making the error so that we can have a isRedirectError(err) function and let the client code handle the call to redirectInfo?
Something like the following (I would find that more readable):
```javascript
api.loginWithMacaroon(new Y.juju.environments.web.Bakery({
  webhandler: new Y.juju.environments.web.WebHandler(),
  interactive: true,
  serviceName: 'juju',
  dischargeStore: window.localStorage
}), err => {
  if (!isRedirectError(err)) {
    // notify the user about the actual error;
    return;
  }
  // We need to redirect the model connection to a new URL.
  api.redirectInfo((servers, err) => {
    if (err) {
      // notify the user about the actual error;
      return;
    }
    // Create a new model connection using one of the "servers".
    ...
  });
});
```